### PR TITLE
MAINT: do not use copyswap in flatiter internals

### LINKED ISSUE #23486: BUG: ndarray.flat is incompatible with new-style dtypes
### Describe the issue:

A segfault is generated from numpy internals when trying to do things with the `flat` iterator.

### Reproduce the code example:

```python
import numpy as np
from numpy.core._multiarray_umath import _get_sfloat_dtype

SF = _get_sfloat_dtype()

scaling = 0.5

x = (np.array([1., 2., 3.])/scaling).view(SF(scaling))

print(x.flat[:])
```

### Error message:

```shell
$ gdb --args $(pyenv which python) setup-scaled-float.py 
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /home/nathan/.pyenv/versions/3.10.9-debug/bin/python...
(gdb) r
Starting program: /home/nathan/.pyenv/versions/3.10.9-debug/bin/python setup-scaled-float.py
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[Detaching after vfork from child process 996063]
[Detaching after vfork from child process 996064]
[Detaching after vfork from child process 996072]
[Detaching after vfork from child process 996073]
[Detaching after vfork from child process 996074]

Program received signal SIGSEGV, Segmentation fault.
0x0000000000000000 in ?? ()
(gdb) up
#1  0x00007ffff668e79d in iter_subscript (self=0x555555a51090, 
    ind=<slice at remote 0x7ffff5cac230>)
    at numpy/core/src/multiarray/iterators.c:634
634	            copyswap(dptr, self->dataptr, 0, ret);
(gdb) p copyswap
$1 = (PyArray_CopySwapFunc *) 0x0
```

### Runtime information:

```
1.25.0.dev0+1036.g84571a1b2
3.10.9 (main, Mar  2 2023, 11:31:32) [GCC 11.3.0]
[{'numpy_version': '1.25.0.dev0+1036.g84571a1b2',
  'python': '3.10.9 (main, Mar  2 2023, 11:31:32) [GCC 11.3.0]',
  'uname': uname_result(system='Linux', node='rous', release='5.19.0-35-generic', version='#36~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Fri Feb 17 15:17:25 UTC 2', machine='x86_64')},
 {'simd_extensions': {'baseline': ['SSE', 'SSE2', 'SSE3'],
                      'found': ['SSSE3',
                                'SSE41',
                                'POPCNT',
                                'SSE42',
                                'AVX',
                                'F16C',
                                'FMA3',
                                'AVX2'],
                      'not_found': ['AVX512F',
                                    'AVX512CD',
                                    'AVX512_KNL',
                                    'AVX512_KNM',
                                    'AVX512_SKX',
                                    'AVX512_CLX',
                                    'AVX512_CNL',
                                    'AVX512_ICL']}}]
```

### Context for the issue:

copyswap isn't defined for any of the new-style dtypes. Sebastian tells me that the fix should be to replace the copyswap with a single-element strided/contiguous copy.

Note also that `array_flat_set` has a similar issue, so that should probably also be fixed when this is, but it's lower priority since that's a less-used code path.

### PR DESCRIPTION
**Pull request by ngoldbaum**
Fixes #23486.

Updates all the code paths used by the flatiter getter and setter to use single-element casts instead of copyswap. This allows new dtypes to use these code paths without seg faulting.

I ran the flatiter benchmarks and didn't see any significant performance changes, but there seem to only be benchmarks for boolean indexing. I could probably substantially improve performance by only using copyswap in situations where it's required but using a hand-unrolled `memcpy` like `fasttake` otherwise but only want to fix the seg faults for now.

### CODE REVIEW

File: numpy/core/src/multiarray/iterators.c [hunk 1]
```diff
@@ -517,23 +532,43 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind)
...
         PyArray_ITER_GOTO1D(self, num);
-        copyswap(optr, self->dataptr, swap, ret);
+        char *args[2] = {self->dataptr, optr};
+        if (cast_info.func(&cast_info.context, args, &one,
+                           transfer_strides, cast_info.auxdata) < 0) {
+            Py_DECREF(ind_it);
+            Py_DECREF(ret);
+            PyArray_ITER_RESET(self);
+            NPY_cast_info_xfree(&cast_info);
+            return NULL;
```
Comment (seberg) [inline numpy/core/src/multiarray/iterators.c:565]: goto might be nice here, but nvm.

File: numpy/core/src/multiarray/iterators.c [hunk 2]
```diff
@@ -501,9 +515,10 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind)
         }
     }
 
-    Py_INCREF(PyArray_DESCR(self->ao));
+    PyArray_Descr *dtype = PyArray_DESCR(self->ao);
+    Py_INCREF(dtype);
     ret = (PyArrayObject *)PyArray_NewFromDescr(Py_TYPE(self->ao),
-                             PyArray_DESCR(self->ao),
+                             dtype,
```
Comment (seberg) [inline numpy/core/src/multiarray/iterators.c:521]: Do you want to check if this works for `"S0"`m OTOH, its hard to actually get an S0 array to begin with.  Maybe the fix with that is to pull the plug and allow "S0" properly.

File: numpy/core/src/multiarray/iterators.c [hunk 3]
```diff
@@ -766,14 +851,21 @@ iter_ass_sub_int(PyArrayIterObject *self, PyArrayObject *ind,
             return -1;
```
Comment (seberg) [inline numpy/core/src/multiarray/iterators.c:851]: Seem to be missing the new cleanup.

File: numpy/core/tests/test_custom_dtypes.py [hunk 4]
```diff
@@ -266,6 +266,25 @@ def test_np_save_load(self):
...
         del np._ScaledFloatTestDType
 
+    def test_flatiter(self):
+        arr = np.array([1.0, 2.0, 3.0], dtype=SF(1.0))
+
+        for i, val in enumerate(arr.flat):
+            assert arr[i] == val
+
+        indices = [[1, 2], ..., slice(None, 2, None),
+                   np.array([True, True, False]), np.array([0, 1])]
```
Comment (seberg) [inline numpy/core/tests/test_custom_dtypes.py:276]: Maybe nice as a parametrization?  (maybe not, fine either way)

File: numpy/core/src/multiarray/iterators.c [hunk 5]
```diff
@@ -627,9 +662,22 @@ iter_subscript(PyArrayIterObject *self, PyObject *ind)
...
         while (n_steps--) {
-            copyswap(dptr, self->dataptr, 0, ret);
+            char *args[2] = {self->dataptr, dptr};
+            if (cast_info.func(&cast_info.context, args, &one,
+                               transfer_strides, cast_info.auxdata) < 0) {
+                goto fail;
+            }
             start += step_size;
             PyArray_ITER_GOTO1D(self, start);
             dptr += size;
```
Comment (seberg) [inline numpy/core/src/multiarray/iterators.c:683]: Cleanup seems also broken in this path, maybe a `goto finish`?  (return 3 lines down)

**Review by seberg — COMMENTED** [review]
Some nits, mostly leave or take (or probably just leave because the below comment makes them not matter anymore).

One larger comment.  It seems to me we can pass in the function into `iter_ass_sub_int`, `iter_ass_sub_Bool`, etc.  Which means we can liberate getting the transfer function to before the `if (PySlice_Check(ind))`.
That will mean we only to the code once rather than 3 times and the cleanup is fully outside of those helpers.

The same should work for the `iter_subscript*` paths, with the added `finish:` label and goto finish.

File: numpy/core/src/multiarray/iterators.c (hunk 2 above)
Comment (ngoldbaum) [reply numpy/core/src/multiarray/iterators.c:509]: Yeah, I guess I'm not sure how to get an SO array:

```
In [2]: arr = np.zeros((3, 3), dtype="S0")

In [3]: arr
Out[3]: 
array([[b'', b'', b''],
       [b'', b'', b''],
       [b'', b'', b'']], dtype='|S1')

In [4]: arr.astype("S0")
Out[4]: 
array([[b'', b'', b''],
       [b'', b'', b''],
       [b'', b'', b'']], dtype='|S1')

In [5]: arr.view("S0")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[5], line 1
----> 1 arr.view("S0")

ValueError: When changing to a smaller dtype, its size must be a divisor of the size of original dtype
```

**Comment by ngoldbaum** [comment]
Thanks for the review @seberg! Simplified so there's only one cast setup that is passed to the boolean and integer indexing helpers. Let me know if you spot any additional issues.

File: numpy/core/src/multiarray/iterators.c (hunk 2 above)
Comment (seberg) [reply numpy/core/src/multiarray/iterators.c:509]: Thanks, let's not worry about it, I guess the only normal way to get it is as a temporary value in `np.asarray([""])`.

**Review by seberg — APPROVED** [review]
Thanks for that and even doing some of the smaller things.  Cleanup now looks all good to me, so lets put it in.

**Pull request merged**